src/request.cc

```diff
@@ -0,0 +1,275 @@
+      } else {
+        return bad;
+      }
+    case header_line_start:
+      if (input == '\r') {
+        m_state = expecting_newline_3;
+        return indeterminate;
+      } else if (!m_headers.empty() && (input == ' ' || input == '\t')) {
+        m_state = header_lws;
+        return indeterminate;
+      } else if (!is_char(input) || is_ctl(input) || is_tspecial(input)) {
+        return bad;
+      } else {
+        m_headers.push_back(std::pair<std::string, std::string>());
+        m_headers.back().first.push_back(input);
+        m_state = header_name;
+        return indeterminate;
+      }
+    case header_lws:
+      if (input == '\r') {
+        m_state = expecting_newline_2;
+        return indeterminate;
+      } else if (input == ' ' || input == '\t') {
+        return indeterminate;
+      } else if (is_ctl(input)) {
+        return bad;
+      } else {
+        m_state = header_value;
+        m_headers.back().second.push_back(input);
+        return indeterminate;
+      }
+    case header_name:
+      if (input == ':') {
+        m_state = space_before_header_value;
+        return indeterminate;
+      } else if (!is_char(input) || is_ctl(input) || is_tspecial(input)) {
+        return bad;
+      } else {
+        m_headers.back().first.push_back(input);
+        return indeterminate;
+      }
+    case space_before_header_value:
+      if (input == ' ') {
+        m_state = header_value;
+        return indeterminate;
+      } else {
+        return bad;
+      }
+    case header_value:
+      if (input == '\r') {
+        m_state = expecting_newline_2;
+        return indeterminate;
+      } else if (is_ctl(input)) {
+        return bad;
+      } else {
+        m_headers.back().second.push_back(input);
+        return indeterminate;
+      }
+    case expecting_newline_2:
+      if (input == '\n') {
+        m_state = header_line_start;
+        return indeterminate;
+      } else {
+        return bad;
+      }
+    case expecting_newline_3:
+      return (input == '\n') ? good : bad;
+    default:
+      return bad;
+  }
+}
+
+bool Request::is_char(int c) {
+  return c >= 0 && c <= 127;
+}
+
+bool Request::is_ctl(int c) {
+  return (c >= 0 && c <= 31) || (c == 127);
+}
+
+bool Request::is_tspecial(int c) {
+  switch (c) {
+    case '(': case ')': case '<': case '>': case '@':
+    case ',': case ';': case ':': case '\\': case '"':
+    case '/': case '[': case ']': case '?': case '=':
+    case '{': case '}': case ' ': case '\t':
+      return true;
+    default:
+      return false;
+  }
+}
+
+bool Request::is_digit(int c) {
+  return c >= '0' && c <= '9';
+}
+
+/* Getter methods for Request */
+
+std::string Request::raw_request() const {
+  return m_raw_request;
+}
+
+std::string Request::method() const {
+  return m_method;
+}
+
+std::string Request::uri() const {
+  return m_uri;
+}
+
+// TODO: Not quite sure what they mean by "version"
+// Right now I just gather HTTP version major and minor
```

Comment:
HTTP version, i.e. "1.1"
so it would be `http_version_major + "." + http_version_minor`